transcom/mymove
# Adding initial implementation and test fixes

## [B-21353](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A1014422)
## [INT PR]()


## Summary

This story introduces adjustments to how the escalated price is calculated based on the contract year. There can be up to 8 total contract years for a contract. Here are the contract years in order:

1. Base Period Year 1
2. Base Period Year 2
3. Base Period Year 3
4. Option Period 1
5. Option Period 2
6. Award Term 1
7. Award Term 2
8. Option Period 3

Calculations needed to be adjusted to compound the escalation factors of each contract year. For example:
```
(A) Given a price for base period year 1
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = Round[p]
4. p = p * quantity

(B) Given a price for base period year 2
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = p * BasePeriodYear2EscalationFactor
4. p = Round[p]
5. p = p * quantity

(C) Given a price for base period year 3
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = p * BasePeriodYear2EscalationFactor
4. p = p * BasePeriodYear3EscalationFactor
5. p = Round[p]
6. p = p * quantity

(E) Given a price for Option Period 1
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = p * BasePeriodYear2EscalationFactor
4. p = p * BasePeriodYear3EscalationFactor
5. p = p * OptionPeriodYear1EscalationFactor
6. p = Round[p]
7. p = p * quantity

(F) Given a price for Option Period 2
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = p * BasePeriodYear2EscalationFactor
4. p = p * BasePeriodYear3EscalationFactor
5. p = p * OptionPeriodYear1EscalationFactor
6. p = p * OptionPeriodYear2EscalationFactor
7. p = Round[p]
8. p = p * quantity

(G) Given a price for Award Term 1
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = p * BasePeriodYear2EscalationFactor
4. p = p * BasePeriodYear3EscalationFactor
5. p = p * OptionPeriodYear1EscalationFactor
6. p = p * OptionPeriodYear2EscalationFactor
7. p = p * AwardTerm1EscalationFactor
8. p = Round[p]
9. p = p * quantity

(H) Given a price for Award Term 2
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = p * BasePeriodYear2EscalationFactor
4. p = p * BasePeriodYear3EscalationFactor
5. p = p * OptionPeriodYear1EscalationFactor
6. p = p * OptionPeriodYear2EscalationFactor
7. p = p * AwardTerm1EscalationFactor
8. p = p * AwardTerm2EscalationFactor
9. p = Round[p]
10. p = p * quantity

(H) Given a price for Option Period 3
1. p = Round[p]
2. p = p * BasePeriodYear1EscalationFactor
3. p = p * BasePeriodYear2EscalationFactor
4. p = p * BasePeriodYear3EscalationFactor
5. p = p * OptionPeriodYear1EscalationFactor
6. p = p * OptionPeriodYear2EscalationFactor
7. p = p * AwardTerm1EscalationFactor
8. p = p * AwardTerm2EscalationFactor
9. p = p * OptionPeriodYear3EscalationFactor
10. p = Round[p]
11. p = p * quantity
```
  

Is there anything you would like reviewers to give additional scrutiny?

[this article](tbd) explains more about the approach used.

## Verification Steps for the Author

These are to be checked by the author.


## Verification Steps for Reviewers

These are to be checked by a reviewer.


### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

You can control what contract year is used by setting the shipments requested pickup date pn shipment creation and setting it in between a start and end date for a specific contract year.

Database table re_contract_years:
![image](https://github.com/user-attachments/assets/8863a972-f550-4d04-b8d8-a6995402268e)

You can change the shipments Requested Pickup Date in the mto_shipments table, through the UI as a TOO by clicking "Edit Shipment" or as a Prime by clicking "Update Shipment"


1. Access MilMove as a customer
2. Create a shipment and use an HHG Shipment and set the Requested Pickup Date between  the start and end date of the re_contract_years that you want to test
3. After creating the shipment approve the shipment as a SC
4. Approve the shipment as a TOO
5. As a Prime add a destination sit service item to the shipment
6. As the TOO approve the service items for the shipment
7. As the prime create a payment request for a some service items within the shipment
8. Go to the move as a TIO and you can view the service items and their prices on the payment request
![image](https://github.com/user-attachments/assets/cba8db42-96ac-4a53-aae0-51c558398f11)
 #13130 

My Scenario:
Requested Pickup Address =  08/08/2025, which puts the contract year at Award Term 2
![image](https://github.com/user-attachments/assets/51d8e264-48c4-4c9a-9b15-6a39ec99445d)

Base price  in Cents = p = 650
Contract Year = Award Term 2
Final Weight = quantity = 3000

This case should follow steps (H)

1. 650 = Round[650]
2. 650 = 650 * (BasePeriodYear1EscalationFactor = 1)
3. 662.61 = 650 * (BasePeriodYear2EscalationFactor = 1.0.194)
4. 675.7959390000001 = 662.61 * (BasePeriodYear3EscalationFactor = 1.0199)
5. 690.0552333129 = 675.7959390000001 * (OptionPeriodYear1EscalationFactor = 1.0211)
6. 704.8224153057961 = 690.0552333129 * (OptionPeriodYear2EscalationFactor = 1.0214)
7. 718.7074168873203 = 704.8224153057961 * (AwardTerm1EscalationFactor = 1.0197)
8. 733.5127896751991 = 718.7074168873203 * (AwardTerm2EscalationFactor = 1.0206)
9. 734 = Round[733.5127896751991]
10. 22020 = 734 * ((3000/100) = 3000.ToCWTFFloat64())
22020 cents = $220.20
![image](https://github.com/user-attachments/assets/b8b45e01-041e-40dd-8f03-c5b83bcd1334)



### Frontend


### Backend


### Database

#### Any new migrations/schema changes:


## Screenshots
